modules/schema/src/main/java/org/apache/ignite/internal/schema/BinaryTupleSchema.java

```diff
@@ -0,0 +1,247 @@
+
+    /** Mask for size of entries in variable-length offset table. */
+    public static final int VARSIZE_MASK = 0x011;
+
+    /** Flag that indicates null map presence. */
+    public static final int NULLMAP_FLAG = 0b100;
+
+    /** Default value for UUID elements. */
+    public static final UUID DEFAULT_UUID = new UUID(0, 0);
+
+    /** Default value for Date elements (Jan 1st 1 BC). */
+    public static final LocalDate DEFAULT_DATE = LocalDate.of(0, 1, 1);
+
+    /** Default value for Time elements (00:00:00). */
+    public static final LocalTime DEFAULT_TIME = LocalTime.of(0, 0);
+
+    /** Default value for DateTime elements (Jan 1st 1 BC, 00:00:00). */
+    public static final LocalDateTime DEFAULT_DATE_TIME = LocalDateTime.of(0, 1, 1, 0, 0);
+
+    /** Default value for Timestamp elements. */
+    public static final Instant DEFAULT_TIMESTAMP = Instant.EPOCH;
+
+    /**
+     * Tuple element description used for tuple parsing and building.
+     *
+     * For binary tuples encoding of values is determined by its basic type and the value itself. Parameters
+     * like precision and scale defined for columns in schema are not taken into account. The only exception
+     * is the Decimal type where the scale parameter is required for decoding.
+     *
+     * To keep things simple we have the scale parameter everywhere but really use it only for Decimals.
+     */
+    public static final class Element {
+        final NativeTypeSpec typeSpec;
+
+        final int decimalScale;
+
+        final boolean nullable;
+
+        public Element(NativeType type, boolean nullable) {
+            typeSpec = type.spec();
+
+            if (typeSpec == NativeTypeSpec.DECIMAL) {
+                DecimalNativeType decimalType = (DecimalNativeType) type;
+                decimalScale = decimalType.scale();
+            } else {
+                decimalScale = 0;
+            }
+
+            this.nullable = nullable;
+        }
+    }
+
+    /** Descriptors of all tuple elements. */
+    private final Element[] elements;
+
+    /** Indicates if the schema contains one or more nullable elements. */
+    private final boolean hasNullables;
+
+    /**
+     * Constructor.
+     *
+     * @param elements Tuple element descriptors.
+     */
+    private BinaryTupleSchema(Element[] elements) {
+        this.elements = elements;
+        this.hasNullables = Arrays.stream(elements).anyMatch(element -> element.nullable);
+    }
+
+    /**
+     * Create a tuple schema based on a range of row columns.
+     *
+     * @param descriptor Row schema.
+     * @param colBegin First columns in the range.
+     * @param colEnd Last column in the range (exclusive).
+     * @return Tuple schema.
+     */
+    private static BinaryTupleSchema createSchema(SchemaDescriptor descriptor, int colBegin, int colEnd) {
+        int numCols = colEnd - colBegin;
+
+        Element[] elements = new Element[numCols];
+
+        for (int i = 0; i < numCols; i++) {
```

Comment:
Do you mean that?
```suggestion
        for (int i = colBegin; i < numCols; i++) {
```

Reply:
Not quite but there was a bug indeed, thanks for pointing out.